Documentation (#316)
pyodide/pyodide-build: docs/explanation/architecture.md

# How pyodide-build Works

> This page explains the internals of pyodide-build — how it turns a regular Python package into a WebAssembly wheel.

Reviewer: ```suggestion
This page explains how pyodide-build cross compiles normal Linux-compatible Python packages into wheels that can be used with Pyodide.
```

## The build pipeline

When you run `pyodide build .`, the following happens:

```
┌────────────────┐

> │ pyodide build  │  CLI entry point

Reviewer: Fix formatting?

        ▼
┌────────────────┐
│  Environment   │  Install xbuildenv + Emscripten SDK (if needed)
│    setup       │  Set up sysconfig, headers, env vars
└───────┬────────┘
        │
        ▼
┌────────────────┐
│   pypa/build   │  Standard PEP 517 build frontend
│                │  Invokes your build backend (setuptools, meson-python, etc.)
└───────┬────────┘
        │  Build backend calls gcc, g++, cmake, meson, cargo...
        ▼
┌────────────────┐
│  pywasmcross   │  Compiler wrapper — intercepts all tool calls
│                │  Redirects to Emscripten (emcc, em++, emar, etc.)
│                │  Filters incompatible flags, adds Wasm flags
└───────┬────────┘
        │
        ▼
┌────────────────┐
│   Emscripten   │  Compiles C/C++ → WebAssembly (.o, .a, .so)
│  (emcc/em++)   │  Links as SIDE_MODULE
└───────┬────────┘
        │
        ▼
┌────────────────┐
│  Wheel output  │  .so files (Wasm) + Python files → .whl
│                │  Tagged: pyemscripten_YYYY_P_wasm32
└────────────────┘
```

## The compiler wrapper (pywasmcross)

The core of pyodide-build is `pywasmcross` — a compiler wrapper that transparently redirects native compiler calls to Emscripten.

### How it works

> When pyodide-build sets up the build environment, it creates symlinks named after common compiler tools:

Reviewer: ```suggestion
When pyodide-build sets up the build environment, it adds overrides for `gcc` and other compiler toolchain tools to the beginning of the path.
```

> | Symlink | Redirects to |

Reviewer: I'm not sure that we should call them symlinks anywhere it's an implementation detail that doesn't matter.
```suggestion
| Tool | override |
```

|---|---|
| `cc`, `gcc` | `emcc` |
| `c++`, `g++` | `em++` |
| `ar` | `emar` |
| `ranlib` | `emranlib` |
| `strip` | `emstrip` |
| `cmake` | `emcmake cmake` (with toolchain flags) |
| `meson` | `meson` (with cross file injected) |
| `cargo` | `cargo` (with Emscripten target) |

These symlinks all point to `pywasmcross.py`. When invoked, pywasmcross:

> 1. Detects which tool it's impersonating (from the symlink name)

Reviewer: ```suggestion
1. Detects which tool it's impersonating from the symlink name
```

> 2. Rewrites the command line for Emscripten

Reviewer: Steps 3 and 4 both seem to be included in this. Maybe redo them as a and b sublist.

Author: Sounds good

> 3. Filters out incompatible flags

Reviewer: ```suggestion
3. Filters out native flags that emcc cannot handle
```

4. Adds WebAssembly-specific flags
5. Executes the real Emscripten tool

When `pyodide build` runs:

> 1. `Makefile.envs` is parsed to set compiler flags, paths, and environment variables

Reviewer: Does `Makefile.envs` matter to users? Maybe leave this out?

Author: Good point. removed

2. The `sysconfig` module is patched to return target-platform values instead of host values
3. `PATH` is modified so pywasmcross symlinks take priority over native compilers
4. Environment variables (`SIDE_MODULE_CFLAGS`, `CMAKE_TOOLCHAIN_FILE`, `MESON_CROSS_FILE`, etc.) are set

## Build isolation

By default, `pyodide build` uses [pypa/build](https://build.pypa.io/) in isolated mode — just like `python -m build`. This means:

1. A temporary virtual environment is created
2. Build dependencies from `[build-system].requires` are installed
3. The build backend is invoked
4. The virtual environment is discarded

The `--no-isolation` flag skips this and uses the current environment, which is useful for complex builds where you need to manage dependencies yourself.

## Further reading

> 

Reviewer: Add pep 783 and 776?
